hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirRenameOp.java

```diff
@@ -470,17 +475,53 @@ static RenameResult unprotectedRenameTo(FSDirectory fsd,
       }
     } finally {
       if (undoRemoveSrc) {
-        tx.restoreSource();
+        tx.restoreSource(srcStoragePolicyCounts);
       }
       if (undoRemoveDst) { // Rename failed - restore dst
-        tx.restoreDst(bsps);
+        tx.restoreDst(bsps, dstStoragePolicyCounts);
       }
     }
     NameNode.stateChangeLog.warn("DIR* FSDirectory.unprotectedRenameTo: " +
         "failed to rename " + src + " to " + dst);
     throw new IOException("rename from " + src + " to " + dst + " failed.");
   }
 
+  /*
+   * Calculate QuotaCounts based on parent directory and storage policy
+   * 1. If the storage policy of src and dst are different,
+   *  calculate the QuotaCounts of src and dst respectively.
+   * 2. If all parent nodes of src and dst are not set with Quota,
+   *  there is no need to calculate QuotaCount.
+   * 3. if parent nodes of src and dst have Quota configured,
+   *  the QuotaCount is calculated once using the storage policy of src.
+   * */
+  private static void computeQuotaCounts(
+          QuotaCounts srcStoragePolicyCounts,
+          QuotaCounts dstStoragePolicyCounts,
+          INodesInPath srcIIP,
+          INodesInPath dstIIP,
+          BlockStoragePolicySuite bsps,
+          RenameOperation tx) {
+    INode dstParent = dstIIP.getINode(-2);
+    INode srcParentNode = FSDirectory.
+            getFirstSetQuotaParentNode(srcIIP);
+    INode srcInode = srcIIP.getLastINode();
+    INode dstParentNode = FSDirectory.
+            getFirstSetQuotaParentNode(dstIIP);
+    byte srcStoragePolicyID = FSDirectory.getStoragePolicyId(srcInode);
+    byte dstStoragePolicyID = FSDirectory.getStoragePolicyId(dstParent);
+    if (srcStoragePolicyID != dstStoragePolicyID) {
+      srcStoragePolicyCounts.add(srcIIP.getLastINode().
+              computeQuotaUsage(bsps));
+      dstStoragePolicyCounts.add(srcIIP.getLastINode()
```

Comment:
IIUC, this result will be used for the next verify and storage used addition/subtraction for src and dst inode, right? But I am confused if it will meet some issues here, given directory /a/b (whose storage policy is HDD), /c/d (whose storage policy is SSD), when rename from /a/b/r1 (let's assume 1GB used) to /c/d/r2, then total HDD storage used will decrease 1GB, and SSD storage used increase 1GB, this will be different with fact?

Reply:
Thanks for finding this problem. If the Inode sets the StoragePolicy we should use the StoragePolicy calculation of the Inode. I will fix it.